src/dataflow/src/render/join/delta_join.rs

```diff
@@ -22,6 +32,132 @@ use crate::operator::CollectionExt;
+pub struct DeltaStagePlan {
+    /// The relation index into which we will look up.
+    lookup_relation: usize,
+    /// The key expressions to use for the stream source.
+    stream_keys: Vec<MirScalarExpr>,
+    /// The key expressions to use for the lookup source.
+    lookup_keys: Vec<MirScalarExpr>,
+    /// The closure to apply to the concatenation of stream and lookup columns.
+    closure: JoinClosure,
+}
+
+impl DeltaJoinPlan {
+    /// Create a new join plan from the required arguments.
+    pub fn create_from(
+        equivalences: &[Vec<MirScalarExpr>],
+        join_orders: &Vec<Vec<(usize, Vec<MirScalarExpr>)>>,
+        input_mapper: JoinInputMapper,
+        map_filter_project: MapFilterProject,
+    ) -> Self {
+        let number_of_inputs = join_orders.len();
+
+        // Create an empty plan, with capacity for the intended number of path plans.
+        let mut join_plan = DeltaJoinPlan {
+            path_plans: Vec::with_capacity(number_of_inputs),
+        };
+
+        // Each source relation will contribute a path to the join plan.
+        for source_relation in 0..number_of_inputs {
+            // Construct initial join build state.
+            // This state will evolves as we build the join dataflow.
+            let mut join_build_state = JoinBuildState::new(
+                input_mapper.global_columns(source_relation),
+                &equivalences,
+                &map_filter_project,
+            );
+
+            // Initial action we can take on the stream before joining.
+            let initial_closure = join_build_state.extract_closure();
+
+            // Sequence of steps to apply.
+            let mut stage_plans = Vec::with_capacity(number_of_inputs - 1);
+
+            // We track the input relations as they are added to the join so we can figure out
+            // which expressions have been bound.
+            let mut bound_inputs = vec![source_relation];
+            // We use the order specified by the implementation.
+            let order = &join_orders[source_relation];
+
+            for (lookup_relation, lookup_keys) in order.iter() {
+                // rebase the intended key to use global column identifiers.
+                let lookup_keys_rebased = lookup_keys
+                    .iter()
+                    .map(|k| input_mapper.map_expr_to_global(k.clone(), *lookup_relation))
+                    .collect::<Vec<_>>();
+
+                // Keys for the incoming updates are determined by locating
+                // the elements of `next_keys` among the existing `columns`.
+                let stream_keys = lookup_keys_rebased
+                    .iter()
+                    .map(|expr| {
+                        let mut bound_expr = input_mapper
+                            .find_bound_expr(expr, &bound_inputs, &join_build_state.equivalences)
+                            .expect("Expression in join plan is not bound at time of use");
```

Comment:
I'm having a hard time figuring out why we should expect the result to be some here. it seems like `stream_keys` is the set of key expressions from our `source_relation` that we want to compare against `lookup_keys` in the `lookup_relation` and i guess it seems like there must be some equivalence that makes it so that all expressions in `lookup_keys` are equivalent to something in `source_relation` but I'm not sure whats enforcing this or if I'm understanding this right. this comment is also not really related to this specific change.

Reply:
I think the point here is that it would be a malformed plan if we are asked to use as keys certain expressions, but cannot find them equated to any terms that have been bound by other columns. There isn't anything in the Rust type system that requires this though, so if a malformed plan has made it this far, we are in trouble and should bail. We could make the method return a `Result` with a `BadPlanError`, but (unless this gets moved up to the coordinator) we would need to do something similarly catastrophic if we found that.

Reply:
Yeah I get that this isn't enforced by the type system and and I think bailing is fine. It's more like - I'm having a hard time understanding why:
> it would be a malformed plan if we are asked to use as keys certain expressions, but cannot find them equated to any terms that have been bound by other columns. 

is true from a logical / thinking about the problem pov. If we had a query like

```
SELECT *
FROM
a_table a, b_table b, c_table c
WHERE
a.id = b.a_id and
c.id = b.c_id
```

or something like that -- is the idea that when we're trying to join `a` and `b` we know we care about `b.a_id` and so we expect that to be equal to some column in `a` ... because we don't do delta joins for cross joins / we only do it for semijoins? or is there some other more fundamental reason I'm missing?

Reply:
The gist is that when we start from `a`, and are about to join in `b`, the only thing we've been told about that is "you should use `a_id` as the key for `b`". We have to go and track down what that should be equated to when starting from `a`, and it happens that we can find that in the form of `a.id`. The join planner could have done that work too, probably, in the same way that it could now do all of this stuff.

Reply:
Ok I think I get it. if we're told that "you should use a_id as the key for b" but then there's nothing that equals that in `a` -- that doesn't actually correspond to a valid sql query that someone could write. and if it was a cross join we would be told "you should use None as the key for b" which is also fine

Reply:
We would probably be told we should use `[]` as the key, but that's the right idea yes. I think more to the point, we don't even know we are looking for anything in `a`, just in the bound set of columns. It's possible for example that we just find `Literal(7)` or something like that. It just needs to be an expression we can form using each row in the `stream` coming in.